book/src/dev/rfcs/0006-stolon.md

```diff
@@ -0,0 +1,399 @@
+part of a mempool membership query (since the mempool is queried to determine
+whether an advertised transaction is already known).
+
+## Isolated, minimally-distinguishable connections
+
+Using Zcash or Bitcoin over Tor is generally inadvisable. The reason is that
+while Tor provides network-level anonymity, the Bitcoin p2p protocol used in
+Zcash leaks an enormous amount of distinguishing information about a node's
+state, with at best half-hearted and ineffective fingerprinting
+countermeasures.
+
+For instance, peer discovery is implemented the `getaddr` message, which
+requests an `addr` message containing a list of peers and their last-seen
+timestamps. The last-seen timestamps record the time the responding node last
+received a message from that peer.  In addition to node fingerprinting, this is
+probably a useful additional source of information for an adversary seeking to
+trace transaction origins.  Zcashd attempts to prevent fingerprinting by
+serving at most one `addr` message per peer connection, but does so
+statelessly, so tearing down and rebuilding the connection defeats this
+prevention.
+
+Building Zcash connections over Tor is therefore only useful if the connections
+can be:
+
+- completely isolated from all other node state, so that peers cannot make
+  queries revealing information about the node;
+
+- constructed with minimal distinguishing information, such as IP addresses,
+  timestamps, etc.
+
+This is not possible with the network implementation in Zcashd but is fairly
+straightforward to implement using `zebra-network`.  The required steps are:
+
+1.  Change the `Handshake` service to allow full configuration of the data sent
+    while constructing a connection;
+
+2.  Provide a new method that constructs an isolated connection:
+
+    ```rust
+    pub async fn connect_isolated(
+        conn: TcpStream,
+        user_agent: String,
+    ) -> Result<BoxService<Request, Response, BoxedStdError>, BoxedStdError>
+    ```
+
+    This method takes an existing `TcpStream` so that it can be used with a
+    pre-constructed Tor circuit. This method should construct a new `Handshake`
+    instance and configure it with fixed default values. The `Handshake`
+    service takes a handler `Service` used to process inbound requests from the
+    remote peer; `connect_isolated` should pass a no-op `service_fn` that
+    ignores requests.
+
+3.  Use `tokio-socks` to construct a `TcpStream` backed by a fresh Tor circuit,
+    and pass it to `connect_isolated`.
+
+This allows Zebra, or any other user of `zebra-network`, to construct anonymous
+connections to Zcash peers. However, there is some amount of work required to
+glue these pieces together.
+
+This glue is implemented in the `stolon` crate, which provides a single method
+for the relay operation:
+
+```rust
+pub async fn send_transaction(
+    tor_addr: SocketAddr,
+    peer_addr: SocketAddr,
+    transaction: Arc<Transaction>,
+) -> Result<(), BoxError>
+```
```

Comment:
Consider also accepting the following address types:

Tor Socks:
* unix socket:
  * https://2019.www.torproject.org/docs/tor-manual.html.en#SocksPort
  * https://github.com/iCepa/Tor.framework#usage

Zcash Peer:
  * Tor onion address
    * https://zcash.readthedocs.io/en/latest/rtd_pages/tor.html#run-a-zcash-hidden-server

Unix socket support will be important for iOS wallets, and other Tor integrations that don't use ports.

Onion Address support requires a specific configuration of `zcashd`, so it is a lower priority. Using an onion address rather than an exit might also have privacy impacts, because they take a different amount of time. So the relay speed of transactions might be distinguishable.

Reply:
Tor addresses: good idea, I guess we should check whether `tokio-socks` can handle unix sockets. If not, we should add that to the queue of changes to try to eventually upstream.

Onion addresses: I guess we should decide if the `zcashd` Tor integration is a good idea before trying to support it.  I haven't looked into it in detail but my worry is that if it's just connecting the big commingled `main.cpp` state machine to an onion service, it may not be much better (and possibly worse) than using an exit.

Reply:
> Tor addresses: good idea, I guess we should check whether tokio-socks can handle unix sockets. If not, we should add that to the queue of changes to try to eventually upstream.

SOCKS5Stream is only implemented for TCPStream right now:
https://docs.rs/tokio-socks/0.3.0/tokio_socks/tcp/struct.Socks5Stream.html

We'd have to implement it for UnixStream:
https://docs.rs/tokio/0.2.22/tokio/net/struct.UnixStream.html

> Onion addresses: I guess we should decide if the zcashd Tor integration is a good idea before trying to support it.

I think it's ok to delay support until we're sure it's actually good for privacy.

Reply:
WDYT of splitting out the transport support into a separate issue, since we might need some changes to upstream?

Reply:
Sounds great!

I don't want to block the design on unix sockets. I opened #1019.

Unix sockets are also worth considering because they are more secure than localhost ports. Unix sockets are files, they are protected by filesystem permissions. (Unlike ports, which can be accessed by any local process.)

This might not matter so much for our threat model, if we are distributing docker containers. But it could be important for systems that run multiple network processes.

Reply:
I didn't open a ticket for the `zcashd` tor integration and submission via onion addresses.

If we decide it's a good idea, we can open a similar ticket to #1019. We'll want to make `peer_addr` generic over `std::net::SocketAddr` (TCP) and onion addresses.